util/hash_map.h

```diff
@@ -64,4 +71,501 @@ class HashMap {
+      rehash(max_size_ << 1);
+    }
+
+    size_t h = hash_fn(value.first);
+    size_t insert_pos = (size_t)-1;
+    size_t pos = h & mask_;
+
+    // Rehash until we get a short distances. This could loop infinitely if we
+    // have a bad hash function.
+    while (true) {
+      pos = h & mask_;
+
+      bool rehashed = false;
+      // Check if we need to rehash due to distance exceeding 4 bits.
+      for (pos = h & mask_;; pos = (pos + 1) & mask_) {
+        if (!info_[pos]) {
+          break;
+        }
+        if (get_dist(info_[pos]) == 15) {
+          rehash(max_size_ << 1);
+          rehashed = true;
+          break;
+        }
+      }
+
+      if (!rehashed) {
+        break;
+      }
+    }
+
+    pos = h & mask_;
+    uint8_t info = set_info(0, h >> 61);
+    // Displace elements
+    while (true) {
+      // If slot is empty, just insert value there.
+      if (!info_[pos]) {
+        if (insert_pos == (size_t)-1) insert_pos = pos;
+        ::new ((void*)&values_[pos]) value_type(std::move(value));
+        info_[pos] = info;
+        break;
+      }
+
+      // If the occupying element has a smaller distance than the inserted
+      // element, then swap the two elements, and continue with inserting the
+      // displaced element.
+      if (get_dist(info_[pos]) < get_dist(info)) {
+        if (insert_pos == (size_t)-1) insert_pos = pos;
+
+        std::swap(value, values_[pos]);
+        std::swap(info, info_[pos]);
+      }
+
+      // Assert that distance will not overflow the 4 bits. We should have
+      // checked in the first loop whether we need to rehash or not due to
+      // distance exceeding 4 bits already.
+      assert(get_dist(info) < 15);
+
+      info = inc_dist(info);
+      pos = (pos + 1) & mask_;
+    }
+
+    size_++;
+    assert(insert_pos != (size_t)-1);
+    check();
+    return std::make_pair(iterator(this, insert_pos), true);
+  }
+
+  ROCKSDB_FORCE_INLINE mapped_type& operator[](key_type const& key) {
+    auto it = find(key);
+    if (it == end()) {
+      return insert({key, mapped_type()}).first->second;
+    }
+    return it->second;
+  }
+
+  ROCKSDB_FORCE_INLINE size_type count(key_type const& key) const {
+    return find(key) == end() ? 0 : 1;
+  }
+
+  ROCKSDB_FORCE_INLINE const_iterator find(const K& key) const {
+    size_t h = hash_fn(key);
+    size_t pos = h & mask_;
+    uint8_t info = set_info(0, h >> 61);
+    while (true) {
+      if (info_[pos]) {
+        if (get_dist(info_[pos]) < get_dist(info)) {
+          break;
+        }
+        if (info_[pos] == info) {
+          if (LIKELY(values_[pos].first == key)) {
+            return const_iterator(this, pos);
+          }
+        }
+      } else {
+        break;
+      }
+      pos = (pos + 1) & mask_;
+      assert(get_dist(info) < 15);
+      info = inc_dist(info);
+    }
+
+    return end();
+  }
+
+  ROCKSDB_FORCE_INLINE iterator find(const K& key) {
+    const_iterator it = const_cast<typename std::add_const<
```

Comment:
I think one way around the constness problems here would be to move the actual find logic to a private helper that would return only an index, and then have two thin find wrappers around it (one const method that returns a const_iterator, and one non-const method that returns an iterator).